packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java

```diff
@@ -0,0 +1,145 @@
+
+  @VisibleForTesting public CameraXProxy cameraXProxy = new CameraXProxy();
+  @VisibleForTesting public TextureRegistry.SurfaceTextureEntry flutterSurfaceTexture;
+
+  public PreviewHostApiImpl(
+      BinaryMessenger binaryMessenger,
+      InstanceManager instanceManager,
+      TextureRegistry textureRegistry) {
+    this.binaryMessenger = binaryMessenger;
+    this.instanceManager = instanceManager;
+    this.textureRegistry = textureRegistry;
+  }
+
+  /** Creates a {@link Preview} with the target rotation and resolution if specified. */
+  @Override
+  public void create(
+      @NonNull Long identifier,
+      @Nullable Long rotation,
+      @Nullable GeneratedCameraXLibrary.ResolutionInfo targetResolution) {
+    Preview.Builder previewBuilder = cameraXProxy.createPreviewBuilder();
+    if (rotation != null) {
+      previewBuilder.setTargetRotation(rotation.intValue());
+    }
+    if (targetResolution != null) {
+      previewBuilder.setTargetResolution(
+          new Size(
+              targetResolution.getWidth().intValue(), targetResolution.getHeight().intValue()));
+    }
+    Preview preview = previewBuilder.build();
+    instanceManager.addDartCreatedInstance(preview, identifier);
+  }
+
+  /**
+   * Sets the {@link Preview.SurfaceProvider} that will be used to provide a {@code Surface} backed
+   * by a Flutter {@link TextureRegistry.SurfaceTextureEntry} used to build the {@link Preview}.
+   */
+  @Override
+  public Long setSurfaceProvider(@NonNull Long identifier) {
+    Preview preview = (Preview) Objects.requireNonNull(instanceManager.getInstance(identifier));
+    flutterSurfaceTexture = textureRegistry.createSurfaceTexture();
+    SurfaceTexture surfaceTexture = flutterSurfaceTexture.surfaceTexture();
+    Preview.SurfaceProvider surfaceProvider = createSurfaceProvider(surfaceTexture);
+    preview.setSurfaceProvider(surfaceProvider);
+
+    return flutterSurfaceTexture.id();
+  }
+
+  /**
+   * Creates a {@link Preview.SurfaceProvider} that specifies how to provide a {@link Surface} to a
+   * {@code Preview} that is backed by a Flutter {@link TextureRegistry.SurfaceTextureEntry}.
+   */
+  @VisibleForTesting
+  public Preview.SurfaceProvider createSurfaceProvider(SurfaceTexture surfaceTexture) {
+    return new Preview.SurfaceProvider() {
+      @Override
+      public void onSurfaceRequested(SurfaceRequest request) {
+        surfaceTexture.setDefaultBufferSize(
+            request.getResolution().getWidth(), request.getResolution().getHeight());
+        Surface flutterSurface = cameraXProxy.createSurface(surfaceTexture);
+        request.provideSurface(
+            flutterSurface,
+            Executors.newSingleThreadExecutor(),
+            (result) -> {
+              SystemServicesFlutterApiImpl systemServicesFlutterApi =
+                  cameraXProxy.createSystemServicesFlutterApiImpl(binaryMessenger);
+              int resultCode = result.getResultCode();
+              switch (resultCode) {
```

Comment:
TLDR: To address a comment @reidbaker left on this logic (that is gone?), I re-read the docs on the various results and updated the code to my current understanding (described in more detail below) that leans more towards releasing the surface when we can, following [this advice from the docs](https://developer.android.com/reference/android/view/Surface#release()). Looking for feedback on the updated `switch` statement. @xizhang 

These are the docs on the results: https://developer.android.com/reference/androidx/camera/core/SurfaceRequest.Result. This is my current understanding that is reflected in the code:

**Error Cases**
- In the case of an invalid surface, the surface could not have been used, so we want to send an error. We shouldn’t have to release it because it is invalid.

**Non-Error Cases**
- In the case of a canceled request, it is safe to release or reuse the surface, so we can release it.
- In the case that the surface was already provided, the request was already completed and it is safe to release or reuse the surface, so we can release it.
- In the case that the surface was used successfully and it is not safe to reuse the surface, so we should release it.
- In the case that a request will not provide a surface, it is safe to release or reuse the surface, so we can release it.

I have it defaulting to an error, but open to feedback on that.

Reply:
We also need to release the Surface when there is an error.

Also FYI `RESULT_INVALID_SURFACE` is not being sent by CameraX right now. But we should handle it nonetheless in case CameraX handles it in the future. The `sendCameraError` part LGTM.

Reply:
Given @xizhang's comment what do you think about this formatting 
```
// See https://developer.android.com/reference/androidx/camera/core/SurfaceRequest.Result for documentation. 
// Always attempt a release.
flutterSurface.release();
int resultCode = result.getResultCode();
switch (resultCode) {
    case SurfaceRequest.Result.RESULT_REQUEST_CANCELLED:
    case SurfaceRequest.Result.RESULT_WILL_NOT_PROVIDE_SURFACE:
    case SurfaceRequest.Result.RESULT_SURFACE_ALREADY_PROVIDED:
    case SurfaceRequest.Result.RESULT_SURFACE_USED_SUCCESSFULLY:
        // Only need to release, do nothing. 
       break;
    case SurfaceRequest.Result.RESULT_INVALID_SURFACE: // Intentional fall through. 
    default: 
        // Release and send error. 
        systemServicesFlutterApi.sendCameraError(getProvideSurfaceErrorDescription(resultCode), reply -> {});
        break;
```

This code will make sure we know that we are missing a newly added case.

Reply:
Love this. Thanks!